core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -252,13 +291,19 @@ public boolean tryComplete() {
                 // replicaManager.readFromLog to populate the offset metadata and update the fetch offset metadata for
                 // those topic partitions.
                 LinkedHashMap<TopicIdPartition, LogReadResult> replicaManagerReadResponse = maybeReadFromLog(topicPartitionData);
+                // Store the remote fetch info and the topic partition for which we need to perform remote fetch.
+                Optional<TopicPartitionRemoteFetchInfo> topicPartitionRemoteFetchInfoOpt = maybePrepareRemoteStorageFetchInfo(topicPartitionData, replicaManagerReadResponse);
+
+                if (topicPartitionRemoteFetchInfoOpt.isPresent()) {
+                    return maybeProcessRemoteFetch(topicPartitionData, topicPartitionRemoteFetchInfoOpt.get());
+                }
                 maybeUpdateFetchOffsetMetadata(topicPartitionData, replicaManagerReadResponse);
                 if (anyPartitionHasLogReadError(replicaManagerReadResponse) || isMinBytesSatisfied(topicPartitionData, partitionMaxBytesStrategy.maxBytes(shareFetch.fetchParams().maxBytes, topicPartitionData.keySet(), topicPartitionData.size()))) {
                     partitionsAcquired = topicPartitionData;
-                    partitionsAlreadyFetched = replicaManagerReadResponse;
+                    localPartitionsAlreadyFetched = replicaManagerReadResponse;
                     boolean completedByMe = forceComplete();
                     // If invocation of forceComplete is not successful, then that means the request is already completed
-                    // hence release the acquired locks.
+                    // hence the acquired locks are already released.
```

Comment:
This comment is still not quite accurate. If `forceComplete()` returns false, it actually means that the locks haven't been released yet.  How about the following?

`If the delayed operation is completed by me, the acquired locks are already released in onComplete(). Otherwise, need to release the acquired locks. `

Also, we call `forceComplete()` in 4 different places and there are quite a bit of duplicated code/comment. Could we introduce a private method and reuse the code in all those places?